AX: Start caching default values for Font and TextColor on the isolated tree

hoffmanjoshua opened the pull request:
#### 9e82aedf3a00f3c3b05538b623c70f8ff915209a
<pre>
AX: Start caching default values for Font and TextColor on the isolated tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=292628">https://bugs.webkit.org/show_bug.cgi?id=292628</a>
<a href="https://rdar.apple.com/150788269">rdar://150788269</a>

Reviewed by Tyler Wilcock.

We are caching TextColor and Font on almost every static text object, since
there is no reasonable predetermined default for all web pages.

To be smarter about this, we can instead cache these properties only when
they differ from their ancestor. This significantly reduces the number of
times we need to cache these properties, since most often these are
inhereted from an ancestor.

I confirmed that these changes don&apos;t have a perf impact by taking a few
samples.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::colorAttributeValue const):
(WebCore::AXIsolatedObject::fontAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::setDefaultTextColor):
(WebCore::AXIsolatedTree::defaultTextColor):
(WebCore::AXIsolatedTree::setDefaultFont):
(WebCore::AXIsolatedTree::defaultFont):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/294634@main">https://commits.webkit.org/294634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6799b2cb7f2b95a0684894929efab3d7110a2545

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22149 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12463 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107641 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53117 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104519 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22457 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30655 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77967 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34959 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17401 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92505 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58303 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17238 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10533 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52474 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10604 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110016 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29612 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21840 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86947 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29976 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88701 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86539 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22027 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31371 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9093 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23862 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29540 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34843 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29351 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32674 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30910 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -609,6 +609,16 @@ void AXIsolatedObject::setProperty(AXProperty property, AXPropertyValueVariant&&
         [](AccessibilityButtonState& typedValue) { return typedValue == AccessibilityButtonState::Off; },
         [&](Color& typedValue) {
+            if (property == AXProperty::TextColor) {
+                auto defaultColor = tree()->defaultTextColor();
```
For std::optional types, I find `std::optional defaultColor` reads better than auto, kinda similar to Ref / RefPtr which we spell out rather than using `auto`. Otherwise it makes me wonder what we're null-checking below — is it a pointer? An optional? Something else? But not deadset on having it this way

hoffmanjoshua replied:
That makes sense to me, I updated this to be more explicit

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -641,7 +651,15 @@ void AXIsolatedObject::setProperty(AXProperty property, AXPropertyValueVariant&&
         [](std::shared_ptr<AXTextRuns> typedValue) { return !typedValue || !typedValue->size(); },
-        [](RetainPtr<CTFontRef>& typedValue) { return !typedValue; },
+        [&](RetainPtr<CTFontRef>& typedValue) {
+            auto defaultFont = tree()->defaultFont();
```
Same comment here about `auto`.

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1126,15 +1144,30 @@ Path AXIsolatedObject::pathAttributeValue(AXProperty property) const
-    if (index == notFound)
+    if (index == notFound) {
+        if (property == AXProperty::TextColor)
+            return tree()->defaultTextColor() ? tree()->defaultTextColor().value() : Color();
```
`AXIsolatedTree::defaultColor` takes a lock, so we probably shouldn't call it twice, once to null-check it. Maybe something like:

```cpp
 if (property == AXProperty::TextColor)
    return tree()->defaultTextColor().value_or(Color());
```

hoffmanjoshua replied:
Ah, I didn't know about `value_or`—this is a lot cleaner

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1167,6 +1167,36 @@ void AXIsolatedTree::setSelectedTextMarkerRange(AXTextMarkerRange&& range)
+    AXTRACE("AXIsolatedTree::setPendingDefaultTextColor"_s);
+    ASSERT(isMainThread());
+
+    Locker locker { m_treeDefaultsLock };
```
I like this name for the lock 👍

minorninth commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -470,6 +472,12 @@ void AXIsolatedObject::initializeProperties(const Ref<AccessibilityObject>& axOb
 
     shrinkPropertiesAfterUpdates();
+
+//    for (auto property : m_properties) {
```
nit: delete this?

minorninth reviewed (approved):
lgtm!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -612,6 +613,9 @@ void AXIsolatedObject::setProperty(AXProperty property, AXPropertyValueVariant&&
         [](AccessibilityButtonState& typedValue) { return typedValue == AccessibilityButtonState::Off; },
         [&](Color& typedValue) {
+            if (property == AXProperty::TextColor)
+                return parentObject() && typedValue == parentObject()->stylesForAttributedString().textColor;
```
Rather than getting `parentObject()` twice, could we store it in a temporary? Same comment for the couple other places we do this. `parentObject()` is not as cheap as you think (looking at a sample right now where it shows up pretty heavy 😅)

twilco replied:
Rather than getting `parentObject()->stylesForAttributedString()` to get the text color, I wonder if it's worth creating a dedicated virtual `textColor` function to just get that. `stylesForAttributedString` requires several other property lookups and creation of a large struct, of which we don't use the result here.

hoffmanjoshua committed:
AX: Start caching default values for Font and TextColor on the isolated tree
https://bugs.webkit.org/show_bug.cgi?id=292628
rdar://150788269

Reviewed by Tyler Wilcock.

We are caching TextColor and Font on almost every static text object, since
there is no reasonable predetermined default for all web pages.

To be smarter about this, we can instead cache these properties only when
they differ from their ancestor. This significantly reduces the number of
times we need to cache these properties, since most often these are
inhereted from an ancestor.

I confirmed that these changes don't have a perf impact by taking a few
samples.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::colorAttributeValue const):
(WebCore::AXIsolatedObject::fontAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::setDefaultTextColor):
(WebCore::AXIsolatedTree::defaultTextColor):
(WebCore::AXIsolatedTree::setDefaultFont):
(WebCore::AXIsolatedTree::defaultFont):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/294634@main
